File: src-tauri/src/app/helpers/factories/lockfile_factory.rs
use crate::app::helpers::factories::ModFactory;
use crate::app::Config;
use std::fs;

#[cfg(test)]
pub struct LockfileFactory {
    mods: Vec<ModFactory>,
}

#[cfg(test)]
impl LockfileFactory {
    pub fn new() -> Self {
        Self {
            mods: Vec::new(),
        }
    }

    pub fn with_mod(mut self, m: &ModFactory) -> Self {
        self.mods.push(m.clone());
        self
    }

    pub fn with_mods(mut self, mods: &Vec<ModFactory>) -> Self {
        self.mods = mods.clone();
        self
    }

    pub fn seed_cache(self) -> Self {
        for m in &self.mods {
            m.seed_cache();
        }
        self
    }

    pub fn seed_mods(self) -> Self {
        for m in &self.mods {
            m.seed_mod();
        }
        self
    }

    pub fn write(self) {
        let mods_json = self
            .mods
            .iter()
            .map(|m| {
                (
                    m.id.clone(),
                    serde_json::json!({
                        "id": m.id,
                        "version": m.version,
                        "minecraft_versions": m.minecraft_versions,
                        "url": m.url,
                        "hash": m.hash(),
                    }),
                )
            })
            .collect::<serde_json::Map<_, _>>();

        let lock = serde_json::json!({ "mods": mods_json });

        fs::write(
            Config::lock_path(),
            serde_json::to_string_pretty(&lock).unwrap(),
        )
            .unwrap();
    }
}

File: packages/core/src/testing/lockfile-factory.ts
import {writeFileSync} from "fs";
import type {ConfigPaths} from "../models/config.js";
import type {ResourceKind} from "../models/manifest.js";
import type {ModFactory} from "./mod-factory.js";

/** Factory for creating test mcpm.lock files */
export class LockfileFactory {
  private byKind: Record<ResourceKind, ModFactory[]> = { mod: [], datapack: [], resourcepack: [], shaderpack: [] };

  private constructor() {}

  static create(): LockfileFactory {
    return new LockfileFactory();
  }

  withMod(m: ModFactory): this {
    this.byKind.mod.push(m);
    return this;
  }

  withMods(mods: ModFactory[]): this {
    this.byKind.mod = [...mods];
    return this;
  }

  withDatapack(d: ModFactory): this {
    this.byKind.datapack.push(d);
    return this;
  }

  withDatapacks(datapacks: ModFactory[]): this {
    this.byKind.datapack = [...datapacks];
    return this;
  }

  withResourcepack(r: ModFactory): this {
    this.byKind.resourcepack.push(r);
    return this;
  }

  withResourcepacks(resourcepacks: ModFactory[]): this {
    this.byKind.resourcepack = [...resourcepacks];
    return this;
  }

  withShaderpack(s: ModFactory): this {
    this.byKind.shaderpack.push(s);
    return this;
  }

  withShaderpacks(shaderpacks: ModFactory[]): this {
    this.byKind.shaderpack = [...shaderpacks];
    return this;
  }

  writeTo(paths: ConfigPaths): void {
    writeFileSync(
      paths.lockPath,
      JSON.stringify(
        {
          mods: toLockObj(this.byKind.mod),
          datapacks: toLockObj(this.byKind.datapack),
          resourcepacks: toLockObj(this.byKind.resourcepack),
          shaderpacks: toLockObj(this.byKind.shaderpack),
        },
        null,
        2,
      ),
    );
  }
}

function toLockObj(items: ModFactory[]): Record<string, unknown> {
  const obj: Record<string, unknown> = {};
  for (const m of items) {
    obj[m.id] = {
      id: m.id,
      version: m.version,
      minecraft_versions: m.minecraftVersions,
      url: m.url,
      hash: m.hash,
    };
  }
  return obj;
}
